File: cap08/src/algoritmos/OrdenacionMezcla.py
def ordenacionMezcla(lista):
    if len(lista) <= 1:
        return lista

    centro = (int)(len(lista) / 2)
    izquierda = []
    derecha = []

    for i in range(centro):
        izquierda.append(lista[i])

    for i in range(centro, len(lista)):
        derecha.append(lista[i])
     
    izquierda = ordenacionMezcla(izquierda)
    derecha = ordenacionMezcla(derecha)
    solucion = mezcla(izquierda, derecha)
    return solucion

def mezcla(izquierda, derecha):
    indexIz = 0
    indexDr = 0
    solucion = []

    while indexIz < len(izquierda) and indexDr < len(derecha):
        if izquierda[indexIz] < derecha[indexDr]:
            solucion.append(izquierda[indexIz])
            indexIz += 1
        else:
            solucion.append(derecha[indexDr])
            indexDr += 1

    while indexIz < len(izquierda):
        solucion.append(izquierda[indexIz])
        indexIz += 1

    while indexDr < len(derecha):
        solucion.append(derecha[indexDr])
        indexDr += 1
    
    return solucion

ejemplo = [7, 5, 4, 1, 9, 8, 2, 3]
ordenada = ordenacionMezcla(ejemplo)
print(ordenada)

File: cap08/src/algoritmos/OrdenacionMezcla.java
package algoritmos;

import java.util.Arrays;

public class OrdenacionMezcla {
	
	public static int[] ordenacionMezcla(int[] lista) {
		if (lista.length <= 1) {
			return lista;
		}
		int centro = lista.length / 2;
		int[] izquierda = Arrays.copyOfRange(lista, 0, centro);
		int[] derecha = Arrays.copyOfRange(lista, centro, lista.length);
		izquierda = ordenacionMezcla(izquierda);
		derecha = ordenacionMezcla(derecha);
		return mezcla(izquierda, derecha);
	}

	private static int[] mezcla(int[] izquierda, int[] derecha) {
		int indexIz = 0;
		int indexDr = 0;
		int indexSol = 0;
		int[] sol = new int[izquierda.length + derecha.length]; 
		while (indexIz < izquierda.length && indexDr < derecha.length) {
			// el siguiente elemento será el más pequeño de cada mitad
			if (izquierda[indexIz] < derecha[indexDr]) {
				sol[indexSol++] = izquierda[indexIz++];
			} else {
				sol[indexSol++] = derecha[indexDr++];
			}
		}
		// copiamos los elementos que faltan de cada cadena
		while (indexIz < izquierda.length) {
			sol[indexSol++] = izquierda[indexIz++];			
		}
		while (indexDr < derecha.length) {
			sol[indexSol++] = derecha[indexDr++];
		}
		return sol;
	}
	
	public static void main(String[] args) {
		System.out.println(Arrays.toString(ordenacionMezcla(
				new int[] { 7, 5, 4, 1, 9, 8, 2, 3 })));
	}
}
